### ISSUE TITLE
Limiting each object monitor to a single reservation

### ISSUE THREAD

**Issue by jdmpapin** [eclipse-openj9/openj9#1124]
I'm looking at a workload in which lock reservation on `StringBuffer` objects causes a significant slowdown, because a relatively small number of them move repeatedly from thread to thread, causing repeated cancellations. I'd like to deal with this by preventing reservation of an object's monitor after it has been cancelled for the first time.

Currently, a reserving monitor entry will happily reserve any unowned and unreserved object (lock word value 0). In order to prevent reservation, the lock word must specify somehow whether or not an available lock is reservable. Specially marking the unreservable locks would require all monitor operations to preserve the distinction, which would probably require an additional flag in the lock word. Marking the *reservable* locks, OTOH, can be done with a much smaller change to the lock word state space.

Say a lock is reservable iff it belongs to a class using reservation, and it hasn't been locked yet. Initialize the lock words of instances of these classes to a previously unused bit pattern that indicates the lock is both available and reservable. The first successful operation on a lock takes that lock, after which it will naturally never return to the reservable bit pattern. This is similar to the reservation strategy put forth in [Kiyokuni Kawachiya's dissertation](http://researcher.watson.ibm.com/researcher/files/jp-KAWATIYA/Kawachiya05phd.pdf). From page 64:

> When an object is created, the lock is *anonymously reserved*. That is, the lock word is in the reserve mode, but not reserved for or held by any particular thread. [...] we use the *initial-locker policy* in our algorithm. [...] When the reservation is canceled, the LRV bit is cleared, and the lockword is put in the base mode. The structure is completely defined by the base algorithm.

Note that with this strategy, if a reservable lock is first taken via a non-reserving path, then it will not be eligible for reservation.

**Comment by jdmpapin** [comment]
It looks like `OBJECT_HEADER_LOCK_RESERVED` is a good choice for the reservable bit pattern. Fast attempts to lock without blocking by running a compare-and-swap against 0 will fail on reservable locks, but they already fail on *reserved* locks, and when they fail the caller attempts `objectMonitorEnterNonBlocking()`. This latter function can be adjusted pretty easily to accept reservable locks near

    /* check to see if object is unlocked (...) */

and to have the compare-and-swap expect `lock` instead of 0. I think that with this change, `objectMonitorEnterBlocking()` never has to deal with reservable locks. If the compare-and-swap succeeded, then the lock is no longer reservable (and there's no need to block). If OTOH it failed, then some other thread has taken the lock, and so it's no longer reservable regardless.

The reservable lock value might be first encountered by operations other than entry, but those will see a flat lock not owned by the current thread.

I've tried a prototype of this on my workload, and it prevents the slowdown that I was seeing. The changes to locking code are [pretty small](https://github.com/eclipse/openj9/files/1704807/reservable.patch.txt). Hopefully the locking sequence change for other platforms will be as straightforward as for x86. For the purpose of measurement I initialized the lock word using `JITHelpers` in the `StringBuffer` constructor.

**Does this seem like a reasonable way to go?** If so, I have some follow-up questions about initializing the lock word, which is looking like the most involved part of this. @DanHeidinga, @gacholio could I bother you for your thoughts, or your suggestions on whose thoughts we want?

Each new object subject to lock reservation has to have its lock word initialized. There are two obvious times to do the initialization: during allocation/default-initialization, or during construction.

Using the constructor is probably the easiest way to make sure that the initialization happens. I doubt we want to instrument the constructors of the affected types though, so it would be some kind of intrinsic call in `Object.<init>`. The JIT would convert it directly into a store, or eliminate it, as appropriate. However, such a call might be a significant overhead in the interpreter, and even in jitted code it might not be straightforward to eliminate redundant zeroing just before the initialization. For that reason I'm leaning toward doing it at allocation time.

Setting the lock word during allocation should minimize the additional interpreter overhead, and also provide an opportunity to initialize the lock word only once. The only downside I'm aware of is that allocation happens in multiple places. We have at least the following:

- the `new` bytecode handler in the interpreter,
- JIT evaluators for `TR::New`,
- initialization of contiguous stack-allocated objects,
- `JITHelpers` fast clone/shallow copy methods,

and there may be others I'm missing. Admittedly it's not absolutely necessary to get every object, but I'd like to avoid accidentally exempting objects from lock reservation.

At which point would folks prefer the initialization code? Is there some other time I've missed? If it's during allocation, then where would be the relevant places outside of the JIT? From a cursory look I've found `MM_ObjectAllocationAPI::inlineAllocateObject` and `MM_JavaObjectAllocationModel::initializeJavaObject`, but maybe those aren't quite right, or maybe I've missed something else.

Finally, regardless of the timing of the initialization, some paths will have to set the lock word conditionally, without static knowledge of the type being initialized, and particularly whether it uses reservation. This information needs to be readily available. Today the JIT determines which types use reservation, but the VM more generally will need to become aware of these types. Can we use a flag on `J9Class`? Should the VM perhaps become the source of truth on this?

**Comment by vijaysun-omr** [comment]
@zl-wang FYI

**Comment by fjeremic** [comment]
> - initialization of contiguous stack-allocated objects,

Shouldn't stack-allocated objects be disallowed to reserve any locks? I mean the thread which owns the stack is implicitly reserving the object anyway since it is constructed on it's stack.

Why does this only have to apply to objects of certain classes (i.e. `StringBuffer`)? Shouldn't this work in a more general way on other object types as well? How often do we envision arbitrary objects being allocated all the time and multiple threads locking on them? Does that happen in the wild? In my mind objects which are typically locked on are long lived, in which case it shouldn't really hurt to try reservation on them and if we fail once and have to cancel then so be it. Thoughts?

Interesting idea though I can definitely see it working. I dislike any `JITHelper` hackery though as this does not seem maintainable at all. What if we choose to absorb OpenJDK's `StringBuffer` for example, which isn't out of the realm of possibilities. Also the locations of the lockwords may change, the user can specify that `StringBuffer` does not have a lockword, etc.

I'd definitely like to hear @vijaysun-omr's and @zl-wang's thoughts on this interesting idea.

**Comment by DanHeidinga** [comment]
> I'd like to deal with this by preventing reservation of an object's monitor after it has been cancelled for the first time.

This is an interesting hypothesis - that reservation is only useful until it's first cancelled.  This implies that locking follows a pattern like: A, A, A, ..., A, B, A, B
Lots of the same thread locking until the second thread locks and afterwards there should be no affinity.

Another possible locking pattern is phases:  A, A, ..., A, B, B, B, ..., B, C,...,C
Where an object moves from thread to thread but "stays" with the thread for some period.

We should instrument the locking code in the VM and run some "interesting" workloads.  The data may illuminate the interesting locking patterns.

Before we can propose solutions, we should look at the locking code as well - there are a number of cases: no lockword, lock nursery, using the backfill slot as a lock, etc.  Each would need to be addressed in this design.

As to implementing the check, the place I would look In the interpreter is the`Object.<init>` send target (and also the `genericReturn` bytecode) that already do magic for finalizable objects:
```
	VMINLINE VM_BytecodeAction
	sendTargetEmptyObjectConstructor(REGISTER_ARGS_LIST)
	{
		j9object_t receiver = (j9object_t)*_sp++;
		_pc += 3;
		VM_VMHelpers::checkIfFinalizeObject(_currentThread, receiver);
		return EXECUTE_BYTECODE;
	}
```

**Comment by zl-wang** [comment]
Interesting idea ... however, it looks like catching the "cold" corner case that String(Buffer) is built by multiple threads (we presumed this should never happen .. the original motivation to lockReserve StringBuffer).  And, it could miss the other cases mentioned by Dan.  e.g.  A, B, B, B, B .... B    I can imagine the object is constructed by thread A, and consumed/used later mostly by thread B.  Which case is more important and more meaningful for real workloads?  I think this question had better be settled first.

**Comment by jdmpapin** [comment]
> Shouldn't stack-allocated objects be disallowed to reserve any locks? I mean the thread which owns the stack is implicitly reserving the object anyway since it is constructed on it's stack.

Objects can be optimistically stack-allocated even if they might escape. If control reaches an escape point, the object is moved to the heap. In this case monitor operations can't be eliminated because we don't know for certain that the object is on the stack.

I think we might also be able to pass stack-allocated objects to methods known not to allow them to escape, although there can't be very many of those under HCR.

> In my mind objects which are typically locked on are long lived, in which case it shouldn't really hurt to try reservation on them and if we fail once and have to cancel then so be it.

Possibly!

> What if we choose to absorb OpenJDK's StringBuffer for example, which isn't out of the realm of possibilities.

I definitely agree that we shouldn't do this by putting a `JITHelpers` call into the `StringBuffer` constructor. That was purely to allow me to try out this scheme without resolving all of the relevant questions.

> I dislike any JITHelper hackery though as this does not seem maintainable at all. [...] Also the locations of the lockwords may change, the user can specify that StringBuffer does not have a lockword, etc.

Code using `Unsafe` or `JITHelpers` can already deal with this variation, e.g. [see here](https://github.com/eclipse/openj9/blob/6fa061ffc2730e3d0b3de1c247581b93b478feae/jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java#L577-L585).

> This is an interesting hypothesis - that reservation is only useful until it's first cancelled. This implies that locking follows a pattern like: A, A, A, ..., A, B, A, B
> Lots of the same thread locking until the second thread locks and afterwards there should be no affinity.
> 
> Another possible locking pattern is phases: A, A, ..., A, B, B, B, ..., B, C,...,C
> Where an object moves from thread to thread but "stays" with the thread for some period.

Yes, a locking pattern like this could definitely benefit from reservation, provided that the phases are sufficiently long on average (in terms of the number of top-level acquires). I don't know how many locking operations are required, but I expect it would take quite a few considering the cost of cancellation. My hypothesis is not that there are no such cases, but rather that most locks that benefit are only ever locked by a single thread (A, ..., A).

> Before we can propose solutions, we should look at the locking code as well - there are a number of cases: no lockword, lock nursery, using the backfill slot as a lock, etc. Each would need to be addressed in this design.

Right now reservation is limited to objects with a lock word. It might be good to lift that restriction as well, but I'm not sure it's related to the prevention of expensive repeated reservation/cancellation cycles.

I'm not familiar with the lock nursery or the use of the backfill slot (or for that matter what the backfill slot even is). Do they interact meaningfully with objects that have a lock word?

> As to implementing the check, the place I would look In the interpreter is theObject.<init> send target (and also the genericReturn bytecode) that already do magic for finalizable objects

Thanks!

**Comment by andrewcraik** [comment]
> I definitely agree that we shouldn't do this by putting a JITHelpers call into the StringBuffer constructor. That was purely to allow me to try out this scheme without resolving all of the relevant questions.

+1

> > This is an interesting hypothesis - that reservation is only useful until it's first cancelled. This implies that locking follows a pattern like: A, A, A, ..., A, B, A, B
Lots of the same thread locking until the second thread locks and afterwards there should be no affinity.
Another possible locking pattern is phases: A, A, ..., A, B, B, B, ..., B, C,...,C
Where an object moves from thread to thread but "stays" with the thread for some period.

> Yes, a locking pattern like this could definitely benefit from reservation, provided that the phases are sufficiently long on average (in terms of the number of top-level acquires). I don't know how many locking operations are required, but I expect it would take quite a few considering the cost of cancellation. My hypothesis is not that there are no such cases, but rather that most locks that benefit are only ever locked by a single thread (A, ..., A).

While the pattern described by @zl-wang could still benefit from reservation the major problem that this issue is trying to address is when long usage on a single thread does not occur. Given that real performance problems have been seen, at least stopping the pathological behavior using an approach like the one proposed by @jdmpapin makes sense to me. If we find a case like the one described by @zl-wang then we can look at how to get the performance there. I guess I'm saying there are many theoretical scenarios but of the two we know of, one has a major perf issue and so we should fix that and address other situations once the major perf issue is fixed IMO.

**Comment by zl-wang** [comment]
my concern is:  it fixed one situation and regressed another.  notice that ABBBBBBB...  pattern is caught by existing code.  see if we can come up with a more versatile solution.

**Comment by andrewcraik** [comment]
@zl-wang  and @fjeremic a few points: this optimization was only enabled very recently so it isn't something that has been depended on for a long period of time, the issue of not catching the ABBBBBB pattern is unfortunate but the performance will not be worse than the state of affairs before this optimization, tracking the lock behaviour is not something we currently have infrastructure to track so detecting the ABBBBB pattern is something we can't currently do. If we built the infrastructure to track it that infrastructure would have overhead that would detract from the benefit of the optimization, whether this is worth it in the long run would require empirical study.

If you have a workload that exhibits this pattern then could you please evaluate the proposed change and see what the impact is for you. If this is an academic concern then I think the real-world problem that has been exposed needs to be addressed quickly due to its magnitude (double digit percentage regression) and we can attempt to re-improve the optimization after that to handle these other cases (should we have examples of workloads where there would be a benefit). I would note that tuning locking is a tricky process and will require careful performance evaluation. The change proposed here is the minimally invasive change required to prevent regressions in the case where reservation will never help (a situation currently unhandled by the optimization - an oversight in its design/implementation).

**Comment by fjeremic** [comment]
I'm ok with us fixing an observed issue in the current implementation as long as it is something clean and maintainable. We currently do have background work to improve lock reservation and to capture a more general use case that is not targeted towards whitelisted classes. Are we able to expose a test of the workload in which the observed behavior is exhibited such that any background or ongoing work in the area of lock reservation and future incoming modifications don't re-regress whatever PR fixes this issue?

Note that lock reservation is currently disabled on Z at least due to a deadlock issue reported on an ILOG workload.

**Comment by DanHeidinga** [comment]
> I'm looking at a workload in which lock reservation on StringBuffer objects causes a significant slowdown, 

What is the workload?  Is this workload representative of real world applications?  a microbench?  another benchmark?

Gathering more data on common locking patterns would make it easy to determine how useful / appropriate this kind of change is to common workloads.

> because a relatively small number of them move repeatedly from thread to thread, causing repeated cancellations.

When are these StringBuffer objects allocated?  Is this a small number of objects that are allocated during the startup phase?  Or are they allocated later?  There are other possible strategies for addressing this issue if the life-cycle of these problematic objects are understood.  i.e.: If these objects are only allocated at startup, don't enable lock reservation until the post startup phase.

> If we built the infrastructure to track it that infrastructure would have overhead that would detract from the benefit of the optimization,

This is a bit of a strawman.  The infrastructure to track this can be done under #ifdef and only enabled when doing this kind of investigation.  If necessary, it could be done as a one-off hacked build.  The overhead in a production build should be zero.

> I think the real-world problem that has been exposed needs to be addressed quickly due to its magnitude (double digit percentage regression) 

What is the workload?  How does it compare to real world workloads?  As you said, tuning locking tends to be a black art which makes me hesitant to make this kind of change (which will make object allocation more expensive by at least another test & branch) without more evidence 1) this locking pattern is common & 2) the proposed change doesn't regress other common workloads.

**Comment by DanHeidinga** [comment]
Any further updates on this?  Otherwise, we can close it for now and resurrect it in the future if necessary.

**Comment by andrewcraik** [comment]
@jdmpapin and I have continued to work on this - things are a lot more complicated than first appeared due to issues with reservable locks being locked by non-reserving paths, non-reserving paths seeing reservable locks etc. The work is continuing, but the fix is not going to be trivial. I don't think closing this is the right thing to do.

The sketch of the plan we have been considering would be to enable the conservative fix discussed here where there is a pathlength overhead and throughput cost, but revert to the aggressive reservation mode if the user specifies -Xlockreservation since the user is opting in at that point.

Work is continuing so I think we should leave this open for now.

**Comment by jdmpapin** [comment]
> What is the workload? Is this workload representative of real world applications? a microbench? another benchmark? [...] How does it compare to real world workloads?

The workload is a real application we test internally at IBM. It runs in an app server, and over all it's not dissimilar to DayTrader. I don't think it's typical to access a single `StringBuffer` from multiple threads, but I also don't think it's so unreasonable that it should incur a double-digit percentage slowdown.

>  There are other possible strategies for addressing this issue if the life-cycle of these problematic objects are understood. i.e.: If these objects are only allocated at startup, don't enable lock reservation until the post startup phase.

This would require each instance to track whether lock reservation is allowed, with something at least vaguely like what I proposed above.

> > If we built the infrastructure to track it that infrastructure would have overhead that would detract from the benefit of the optimization,
> 
> This is a bit of a strawman. The infrastructure to track this can be done under #ifdef and only enabled when doing this kind of investigation. If necessary, it could be done as a one-off hacked build. The overhead in a production build should be zero.

I think @andrewcraik was talking about any further tracking that might be done in an attempt to make the performance of lock reservation more robust in the face of different reasonable locking patterns, rather than debug instrumentation to help identify those locking patterns. This would have been in response to @zl-wang's request to find a more versatile solution.

>  As you said, tuning locking tends to be a black art which makes me hesitant to make this kind of change (which will make object allocation more expensive by at least another test & branch) without more evidence 1) this locking pattern is common & 2) the proposed change doesn't regress other common workloads.

These are important concerns. It's been difficult trying to tweak the reservation logic in such a way that reservation is still a net win. I thought that the dissertation I linked when opening this issue was good evidence that it's common for an object to only ever be locked by a single thread. However, it was written at a time when there would have been much heavier use of `StringBuffer`, `Vector`, `Hashtable`, etc., whose synchronization was often completely unnecessary.

**Comment by jdmpapin** [comment]
This was implemented in #2344
